preprocessing/structs.ts

```diff
@@ -96,121 +161,159 @@ const id_field = (name: string) => {
-     * an offer insertion consumes an unbounded amount of gas (because it has to be correctly placed in the book).
-
-Note: An optimization in the `marketOrder` function relies on reentrancy being forbidden.
-     */
-    { name: "lock", bits: 1, type: "bool" },
-    /* * `best` holds the current best offer id. Has size of an id field. *Danger*: reading best inside a lock may give you a stale value. */
-    id_field("best"),
-    /* * `last` is a counter for offer ids, incremented every time a new offer is created. It can't go above $2^{32}-1$. */
-    id_field("last"),
-  ],
+      fields.gasreq,
+      /*
+        * <a id="structs.js/gasbase"></a>  `offer_gasbase` represents the gas overhead used by processing the offer inside Mangrove + the overhead of initiating an entire order, in 1k gas increments.
+
+      The gas considered 'used' by an offer is the sum of
+      * gas consumed during the call to the offer
+      * `offer_gasbase`
+      
+    (There is an inefficiency here. The overhead could be split into an "offer-local overhead" and a "general overhead". That general overhead gas penalty could be spread between all offers executed during an order, or all failing offers. It would still be possible for a cleaner to execute a failing offer alone and make them pay the entire general gas overhead. For the sake of simplicity we keep only one "offer overhead" value.)
+
+    If an offer fails, `gasprice` mwei is taken from the
+    provision per unit of gas used. `gasprice` should approximate the average gas
+    ratio at offer creation time.
+
+    `kilo_offer_gasbase` is the actual field name, and is _9 bits wide_ and represents 1k gas increments. The accessor `offer_gasbase` returns `kilo_offer_gasbase * 1e3`.
+
+    `kilo_offer_gasbase` is also the name of a local Mangrove
+    parameters. When an offer is created, their current value is copied from Mangrove local configuration. The maker does not choose it.
+
+    So, when an offer is created, the maker is asked to provision the
+    following amount of wei:
+    ```
+    (gasreq + offer_gasbase) * gasprice * 1e6
+    ```
+
+      where `offer_gasbase` and `gasprice` are Mangrove's current configuration values (or a higher value for `gasprice` if specified by the maker).
+
+
+      When an offer fails, the following amount is given to the taker as compensation:
+    ```
+    (gasused + offer_gasbase) * gasprice * 1e6
+    ```
+
+    where `offer_gasbase` and `gasprice` are Mangrove's current configuration values.  The rest is given back to the maker.
+
+      */
+      fields.kilo_offer_gasbase,
+      /* * `gasprice` is in mwei/gas and _26 bits wide_, which accommodates 0.001 to ~67k gwei / gas.  `gasprice` is also the name of a global Mangrove parameter. When an offer is created, the offer's `gasprice` is set to the max of the user-specified `gasprice` and Mangrove's global `gasprice`. */
```

Comment:
hmm, unit prefixes (m and g) are a bit confusing here, since they're used in the SI sense. But "gwei" seems often used instead of "Gwei" (though that's also in use according to google).

I'd prefer to stick to SI prefixes to avoid confusion.

```suggestion
      /* * `gasprice` is in Mwei/gas and _26 bits wide_, which accommodates 0.001 to ~67k Gwei / gas.  `gasprice` is also the name of a global Mangrove parameter. When an offer is created, the offer's `gasprice` is set to the max of the user-specified `gasprice` and Mangrove's global `gasprice`. */
```